registration/include/pcl/registration/impl/gicp.hpp

```diff
@@ -343,6 +342,31 @@ GeneralizedIterativeClosestPoint<PointSource, PointTarget>::OptimizationFunctorW
   gicp_->computeRDerivative(x, R, g);
 }
 
+template <typename PointSource, typename PointTarget> inline BFGSSpace::Status
+GeneralizedIterativeClosestPoint<PointSource, PointTarget>::OptimizationFunctorWithIndices::checkGradient(const Vector6d& g)
+{
+  auto linear_epsilon = gicp_->linear_gradient_tolerance_;
+  auto angular_epsilon = gicp_->angular_gradient_tolerance_;
+
+  if ((linear_epsilon < 0.) || (angular_epsilon < 0.))
+    return BFGSSpace::NegativeGradientEpsilon;
+
+  // express linear gradient as norm of translation parameters
+  auto linear_grad = g.head(3).norm();
+
+  // express angular gradient as a rotation around a given axis of a certain positive angle,
+  // we're literally only interested in the magnitude of the angular error.
+  // !!! CAUTION Stanford GICP uses the Z Y X euler angles convention
+  auto R = Eigen::Matrix3d{	Eigen::AngleAxisd(g[5], Eigen::Vector3d::UnitZ()) *
+							Eigen::AngleAxisd(g[4], Eigen::Vector3d::UnitY()) *
+							Eigen::AngleAxisd(g[3], Eigen::Vector3d::UnitX())};
+  auto angular_grad = std::acos(std::min(std::max(0.5 * (R.trace() - 1.), -1.), 1.));
```

Comment:
I must admit I'm a little confused here, because I was trying to follow your advice:

https://github.com/PointCloudLibrary/pcl/issues/3813#issuecomment-607203611

Reply:
Back in that comment, your question was "how to accurately quantify an angular error?".  My answer had that question in mind

This function is not about measuring angular error. It's an optimizer evaluating if it should stop iterating by looking at magnitude of the gradient. This magnitude is known to zero in the optimum. You do so, by looking at the norm of the gradient.

Reply:
I agree, but I find it mathematically odd to take the gradient of a vector composed of euler angles (compared to the norm of a translation vector).

Reply:
I also agree with that. But that's why you added the separate norm checks with separate thresholds, one for the translation components and another for the angular components.

Reply:
@SergioRAgostinho : thanks for the detailed explanation. Apart from a more ambitious optimization redesign, should I stick with your initial proposition for now?

Reply:
> Apart from a more ambitious optimization redesign, should I stick with your initial proposition for now?

Completely, the exponential map thing was just a comment, not an implementation request.